cro-clib/src/transaction_build.rs

```diff
@@ -0,0 +1,294 @@
+    }
+}
+
+/// add txin in bytes
+#[no_mangle]
+/// # Safety
+pub unsafe extern "C" fn cro_tx_add_txin_raw(
+    tx_ptr: CroTxPtr,
+    txid: [u8; 32],
+    txindex: u16,
+    addr: [u8; 32],
+    coin: u64,
+) -> CroResult {
+    let tx = tx_ptr.as_mut().expect("get tx");
+    let txin_pointer = TxoPointer::new(txid, txindex as usize);
+    let txin = TxOut::new(
+        ExtendedAddr::OrTree(addr),
+        Coin::new(coin).expect("get coin in cro_tx_add_txin"),
+    );
+
+    tx.txin_pointer.push(txin_pointer);
+    tx.txin_witness.push(None);
+    tx.txin.push(txin);
+    CroResult::success()
+}
+
+/// add viewkey
+#[no_mangle]
+/// # Safety
+pub unsafe extern "C" fn cro_tx_add_viewkey(
+    tx_ptr: CroTxPtr,
+    viewkey_string: *const c_char,
+) -> CroResult {
+    let tx = tx_ptr.as_mut().expect("get tx");
+    let viewkey = get_string(viewkey_string);
+    let hex: Vec<u8>;
+    if let Ok(value) = hex::decode(&viewkey) {
+        hex = value;
+    } else {
+        return CroResult::fail();
+    }
+    assert!(33 == hex.len());
+    let pubkey: secp256k1::PublicKey = secp256k1::PublicKey::from_slice(&hex[..]).unwrap();
+    tx.viewkey.push(pubkey);
+    CroResult::success()
+}
+
+/// add viewkey in bytes
+#[no_mangle]
+/// # Safety
+pub unsafe extern "C" fn cro_tx_add_viewkey_raw(tx_ptr: CroTxPtr, viewkey: [u8; 33]) -> CroResult {
+    let tx = tx_ptr.as_mut().expect("get tx");
+    let pubkey: secp256k1::PublicKey = secp256k1::PublicKey::from_slice(&viewkey).unwrap();
+    tx.viewkey.push(pubkey);
+    CroResult::success()
+}
+
+fn do_cro_tx_prepare_for_signing(user_tx: &mut CroTx, network: u8) -> Result<()> {
+    let mut tx: Tx = Tx::default();
+    tx.inputs = user_tx.txin_pointer.clone();
+    tx.outputs = user_tx.txout.clone();
+    let mut access_policies = BTreeSet::new();
+    for viewkey_user in &user_tx.viewkey {
+        access_policies.insert(TxAccessPolicy {
+            view_key: viewkey_user.clone(),
+            access: TxAccess::AllData,
+        });
+    }
+    tx.attributes = TxAttributes::new_with_access(network, access_policies.into_iter().collect());
+    user_tx.tx = Some(tx);
+    Ok(())
+}
+
+/// prepare tx for signing
+#[no_mangle]
+/// # Safety
+pub unsafe extern "C" fn cro_tx_prepare_for_signing(tx_ptr: CroTxPtr, network: u8) -> CroResult {
+    let mut user_tx: &mut CroTx = tx_ptr.as_mut().expect("get tx");
+    match do_cro_tx_prepare_for_signing(&mut user_tx, network) {
+        Ok(_) => CroResult::success(),
+        Err(_) => CroResult::fail(),
+    }
+}
+
+/// extract bytes from singed tx
+#[no_mangle]
+/// # Safety
+pub unsafe extern "C" fn cro_tx_complete_signing(
+    tx_ptr: CroTxPtr,
+    output: *mut u8,
+    output_length: *mut u32,
+) -> CroResult {
+    let user_tx: &mut CroTx = tx_ptr.as_mut().expect("get tx");
+    let encoded: Vec<u8> = user_tx.tx.as_ref().expect("get tx core").encode();
+    ptr::copy_nonoverlapping(encoded.as_ptr(), output, encoded.len());
```

Comment:
this doesn't extract the signature though -- perhaps construct PlainTxAux and serialise that?

Reply:
ok